plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkPublishOperation.java

```diff
@@ -0,0 +1,330 @@
+
+  /**
+   * Throws new {@link CoreException} if status list is not empty.
+   */
+  private static void checkStatuses(List<IStatus> statusList) throws CoreException {
+    if (statusList == null || statusList.size() == 0) {
+      return;
+    }
+    if (statusList.size() == 1) {
+      throw new CoreException(statusList.get(0));
+    }
+    IStatus[] children = statusList.toArray(new IStatus[statusList.size()]);
+    throw new CoreException(
+        new MultiStatus(PLUGIN_ID, 0, children,
+        "Error during publish operation", null));
+  }
+
+  private CloudSdkServerBehaviour server;
+  private IModule[] modules;
+  private int kind;
+  private int deltaKind;
+  private PublishHelper helper;
+
+  public int getKind() {
+    return REQUIRED;
+  }
+
+  @Override
+  public int getOrder() {
+    // Hmm, shouldn't the publish tasks come later?
+    return 0;
+  }
+
+  /**
+   * Construct the operation object to publish the specified modules(s) to the specified server.
+   */
+  public CloudSdkPublishOperation(CloudSdkServerBehaviour server, int kind, IModule[] modules,
+      int deltaKind) {
+    super("Publish to server", "Publish a modules to App Engine Development Server");
+    this.server = server;
+    this.modules = modules;
+    this.kind = kind;
+    this.deltaKind = deltaKind;
+    IPath base = server.getRuntimeBaseDirectory();
+    helper = new PublishHelper(base.toFile());
+  }
+
+  @Override
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    // TODO(amitin): use more advanced key to store modules publish locations? Because a dependent
+    // java project (added as child modules and published as jar) cannot present in more than one
+    // parent modules.
+    List<IStatus> statusList = Lists.newArrayList();
+    if (modules.length == 1) {
+      IPath deployPath = server.getModuleDeployDirectory(modules[0]);
+      // root modules
+      publishDir(deployPath, statusList, monitor);
+    } else {
+      // BSD: NEED TO UNDERSTAND SITUATION FOR REMAINING CODE
+      for (int i = 0; i < modules.length - 1; i++) {
+        IPath deployPath = server.getModuleDeployDirectory(modules[i]);
+        IWebModule webModule = (IWebModule) modules[i].loadAdapter(IWebModule.class, monitor);
+        if (webModule == null) {
+          statusList.add(newErrorStatus("Not a Web modules: " + modules[i].getName()));
+          return;
+        }
+        String uri = webModule.getURI(modules[i + 1]);
+        if (uri != null) {
+          deployPath = deployPath.append(uri);
+        } else {
+          // no uri is OK for removed modules
+          if (deltaKind != ServerBehaviourDelegate.REMOVED) {
+            statusList
+                .add(newErrorStatus("Cannot get URI for modules: " + modules[i + 1].getName()));
+            return;
+          }
+        }
+      }
+      // BSD: Disabled saving and loading module locations
```

Comment:
The original code in GPE wrote out these modules to a separate file; I haven't copied those over.  I wonder why they didn't use server attributes instead.